MicroProfile Language Support for Open Liberty blog post (#783)
OpenLiberty/blogs: posts/2020-09-30-microprofile-language-support-liberty.adoc

---
layout: post
title: "MicroProfile Language Support for Open Liberty Applications"
categories: blog
author_picture: https://avatars3.githubusercontent.com/kathrynkodama
author_github: https://github.com/kathrynkodama
seo-title: MicroProfile Language Support for Open Liberty Applications - OpenLiberty.io
seo-description: Leverage the new MicroProfile Language Server along with existing VS Code tools to develop your Eclipse MicroProfile applications with Open Liberty.
blog_description: "Leverage the new MicroProfile Language Server along with existing VS Code tools to develop your Eclipse MicroProfile applications with Open Liberty."
---
= MicroProfile Language Support for Open Liberty Applications
Kathryn Kodama https://github.com/kathrynkodama

> The new MicroProfile Language Server provides added support for developing your Eclipse MicroProfile applications with Open Liberty. Earlier this month, the first release of the Eclipse MicroProfile Language Server was published as a joint effort between IBM and Red Hat. The Tools for MicroProfile VS Code extension consumes this language server, providing feature-rich support for MicroProfile APIs and properties in VS Code.

Reviewer: ```suggestion
The new MicroProfile Language Server provides added support for developing your Eclipse MicroProfile applications with Open Liberty. Earlier this month, the first release of the Language Server for Eclipse MicroProfile (Eclipse LSP4MP), which is an open incubator project under the Eclipse Foundation, was published. The Tools for MicroProfile VS Code extension consumes this language server, providing feature-rich support for MicroProfile APIs and properties in VS Code.
```

Reviewer: If we want to note the participants of the Eclipse project (e.g. IBM, RH, the MP community), we can still.  Then, we will note them under the context of the Eclipse project.

@kathrynkodama @lauracowen ^^^

Author: Updated with the wording you suggested.  I don't feel we need to note the participants, just that it is an open incubator project under the Eclipse Foundation is sufficient.

> In this blog, we will showcase how the new MicroProfile Language Server, along with existing tooling for VS code, makes it even easier to develop your Open Liberty applications.

Reviewer: ```suggestion
In this blog, we will showcase how the new MicroProfile Language Server, along with existing tooling for VS Code, makes it even easier to develop your Open Liberty applications.
```

== Installing the extensions

The following extensions are available on the VS Code Marketplace to support developing your Open Liberty applications with MicroProfile.

- https://marketplace.visualstudio.com/items?itemName=redhat.vscode-microprofile[Tools for MicroProfile]: Language support for MicroProfile APIs and properties via the https://github.com/eclipse/lsp4mp[Language Server for Eclipse].

- https://marketplace.visualstudio.com/items?itemName=Open-Liberty.liberty-dev-vscode-ext[Open Liberty Tools]: Support for development with Open Liberty.

- https://marketplace.visualstudio.com/items?itemName=MicroProfile-Community.mp-starter-vscode-ext[MicroProfile Starter]: Quickly generate a MicroProfile Maven project.

== Setting up the project

=== Create a new MicroProfile project

> Generate a new Open Liberty Maven application with the MicroProfile Starter.

Reviewer 2: Maybe we want to include some quick information on how to use the MP Starter extension as people may not know how.

Ie. Open Command Pallet (Cmd + Shift + P) -> `MicroProfile: Generate a new MicroProfile Starter Project` -> Enter package name + artifact id -> Select MP 3.3 -> Select Open Liberty -> Select Java Version -> Select the specs you want to use.

Or maybe we could include a gif of project generation?

    <artifactId>liberty-maven-plugin</artifactId>
    <version>3.3-M3</version>
</plugin>
----

Add the MicroProfile dependency to your `pom.xml` file.

[source, xml]
----
<dependency>
    <groupId>org.eclipse.microprofile</groupId>
    <artifactId>microprofile</artifactId>
    <version>3.3</version>
    <type>pom</type>
    <scope>provided</scope>
</dependency>
----

For Gradle projects, add the Liberty Gradle Plugin to your `build.gradle` file.

[source,groovy]
----
apply plugin: 'liberty'

buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'io.openliberty.tools:liberty-gradle-plugin:3.1-M2'
    }
}
----

Add the MicroProfile dependency to your `build.gradle` file.

[source,groovy]
----
dependencies {
    providedCompile 'org.eclipse.microprofile:microprofile:3.2'
}
----

Add the MicroProfile features that you would like to use in your Open Liberty `server.xml` file.

[source, xml]
----
<featureManager>
    <feature>jaxrs-2.1</feature>
    <feature>jsonp-1.1</feature>
    <feature>cdi-2.0</feature>
    <feature>mpMetrics-2.0</feature>
    <feature>mpConfig-1.3</feature>
    <feature>mpHealth-2.2</feature>

> </featureManager>

Reviewer 2: Do we recommend people enable features individually as needed, or do we recommend users just use the microprofile convenience features. @yeekangc any thoughts on this? 
```
<featureManager>
  <feature>microProfile-3.3</feature>
<featureManager>
```

Reviewer: It depends on what the users want to accomplish and what we want to accomplish here.  Individual features will give you optimal footprint.

For our purpose here, it's probably alright to use the convenience feature for brevity and simplicity (with a note about that).

Author: Updated to:
```
Add the MicroProfile features you would like to use in your Open Liberty `server.xml` file. 
For convenience, we have specified `microprofile-3.3` to include all of the MicroProfile features.

[source, xml]
----
<featureManager>
    <feature>microProfile-3.3</feature>
</featureManager>
----
```

----

== Starting dev mode

> Start your Open Liberty application in dev mode using the Liberty Dev Dashboard, or via command line using `mvn liberty:dev`/`gradle libertyDev`. Dev mode provides hot reload and deployment, so your code changes will automatically be reflected in your running application.

Reviewer 2: Should we mention liberty dev dashboard is part of vscode? maybe something like:  

Start your Open Liberty application in dev mode using vscode with the Liberty Dev Dashboard, or via command line using?

[.img_border_light]
image::/img/blog/liberty-dev-dashboard-vscode.png[Liberty Dev Dashboard in VS Code]

== Developing with MicroProfile language support

The MicroProfile Language Server provides useful features such as snippets, completion, hover, and diagnostics for both Java and properties files.

=== Coding assistance with Java source files

> Use a snippet to quickly insert popular code blocks. For example,  the  `mpreadiness` snippet inserts a fully functional MicroProfile Health Check Readiness class. Open Liberty dev mode will automatically pick up any changes and recompile your source code so that you can immediately access your new health class at the `/health` endpoint.

Reviewer 2: I think it is just "MicroProfile Health" so maybe change to "inserts a fully functional MicroProfile Health Readiness check class"

[.img_border_light]
image::/img/blog/mp-readiness-snippet.gif[MicroProfile readiness snippet]

> The MicroProfile Language Server has knowledge of the MicroProfile APIs, meaning your source code will validate and correct improper use of MicroProfile features. For example, as per the MicroProfile Health specification, all MicroProfile Health classes using the `@Readiness` or `@Liveness` annotations must implement the `HealthCheck` class. The MicroProfile Language Server will correct you where needed with a helpful diagnostic.

Reviewer 2: Is this better? 

The MicroProfile Language Server has knowledge of the MicroProfile APIs, meaning the language server will validate and correct improper use of MicroProfile APIs in your source code.

---
layout: post
title: "MicroProfile Language Support for Open Liberty Applications"
categories: blog
author_picture: https://avatars3.githubusercontent.com/kathrynkodama
author_github: https://github.com/kathrynkodama
seo-title: MicroProfile Language Support for Open Liberty Applications - OpenLiberty.io

> seo-description: Leverage the new MicroProfile Language Server along with existing VS Code tools to develop your Eclipse MicroProfile applications with Open Liberty.

Reviewer: ```suggestion
seo-description: Leverage the new MicroProfile Language Server along with Visual Studio Code tools to develop your cloud-native Java applications with Eclipse MicroProfile API and Open Liberty.
```

Author: Updated seo-description

> blog_description: "Leverage the new MicroProfile Language Server along with existing VS Code tools to develop your Eclipse MicroProfile applications with Open Liberty."

Reviewer: ```suggestion
blog_description: "Leverage the new MicroProfile Language Server along with VS Code tools to develop your Eclipse MicroProfile applications with Open Liberty."
```

Author: Updated blog description

In this blog, we will showcase how the new MicroProfile Language Server, along with existing tooling for VS Code, makes it even easier to develop your Open Liberty applications.

== Installing the extensions

The following extensions are available on the VS Code Marketplace to support developing your Open Liberty applications with MicroProfile.

- https://marketplace.visualstudio.com/items?itemName=redhat.vscode-microprofile[Tools for MicroProfile]: Language support for MicroProfile APIs and properties via the https://github.com/eclipse/lsp4mp[Language Server for Eclipse].

- https://marketplace.visualstudio.com/items?itemName=Open-Liberty.liberty-dev-vscode-ext[Open Liberty Tools]: Support for development with Open Liberty.

- https://marketplace.visualstudio.com/items?itemName=MicroProfile-Community.mp-starter-vscode-ext[MicroProfile Starter]: Quickly generate a MicroProfile Maven project.

> 

Reviewer: Offer an option for users to install all in one go with MicroProfile Extension Pack?

Author: Updated to: 

Install the https://marketplace.visualstudio.com/items?itemName=MicroProfile-Community.vscode-microprofile-pack[MicroProfile extension pack] from the VS Code Marketplace to support developing your Open Liberty applications with MicroProfile. We will use the following extensions from the MicroProfile extension pack:

[.img_border_light]
image::/img/blog/mp-starter-vscode-marketplace.png[MicroProfile Starter for VS Code]

=== Configure an existing project

Modify an existing Maven or Gradle project for use with Open Liberty and MicroProfile.

For Maven projects, add the Liberty Maven Plugin to your `pom.xml` file.

[source,xml]
----
<plugin>
    <groupId>io.openliberty.tools</groupId>
    <artifactId>liberty-maven-plugin</artifactId>
    <version>3.3-M3</version>
</plugin>
----

Add the MicroProfile dependency to your `pom.xml` file.

[source, xml]
----
<dependency>
    <groupId>org.eclipse.microprofile</groupId>
    <artifactId>microprofile</artifactId>
    <version>3.3</version>
    <type>pom</type>
    <scope>provided</scope>
</dependency>
----

For Gradle projects, add the Liberty Gradle Plugin to your `build.gradle` file.

[source,groovy]
----
apply plugin: 'liberty'

buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'io.openliberty.tools:liberty-gradle-plugin:3.1-M2'
    }
}
----

Add the MicroProfile dependency to your `build.gradle` file.

[source,groovy]
----
dependencies {

>     providedCompile 'org.eclipse.microprofile:microprofile:3.2'

Reviewer: ```suggestion
    providedCompile 'org.eclipse.microprofile:microprofile:3.3'
```
